app/utils/users.ts

```diff
@@ -0,0 +1,95 @@
+"use client";
+import useSWRInfinite from "swr/infinite";
+import { useEffect, useState, useMemo } from "react";
+import { useAuth0 } from "@auth0/auth0-react";
+import { ListUsersPage } from "@/types/users";
+
+const PAGE_SIZE = 50;
+const INITIAL_PAGE_LIMIT = 10;
+
+/**
+ * Basic fetch wrapper that uses a bearer token and
+ * throws an error if something goes wrong
+ */
+const fetcher = async (url: string, token: string) =>
+  fetch(url, {
+    headers: { Authorization: `Bearer ${token}` },
+  }).then((res) => {
+    if (!res.ok) throw new Error("Failed to fetch");
+    return res.json();
+  });
+
+/**
+ * SWR hook that fetches all users up to the first
+ * <INITIAL_PAGE_LIMIT> * <PAGE_SIZE>
+ */
+export function useUsersInfinite(token: string) {
+  const getKey = (pageIndex: number, previousPageData: ListUsersPage) => {
+    if (!token) {
+      return null;
+    }
+    if (previousPageData && previousPageData.users.length < PAGE_SIZE) {
+      // end of data - stop trying to fetch
+      return null;
+    }
+    // return the next URL to fetch
+    return `${process.env.NEXT_PUBLIC_CR3_API_DOMAIN}/user/list_users?page=${pageIndex}&per_page=${PAGE_SIZE}`;
+  };
+  const { data: pages, isLoading } = useSWRInfinite<ListUsersPage>(
```

Comment:
You can see we're passing the `ListUsersPage` type into this hook, and assuming the API is honoring its part of the contract. We can set up zod validation here if we want to.